src/Microsoft.ML.Transforms/TensorFlow/TensorGeneric.cs

```diff
@@ -0,0 +1,140 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+
+namespace Microsoft.ML.Transforms.TensorFlow
+{
+    internal partial class TFTensor
+    {
+        /// <summary>
+        /// Creates a tensor representing type T.
+        /// If T is a scalar the tensor will be backed with a managed-heap-allocated T.
+        /// If T is a single or multi-dimensional array, that array will be pinned and wrapped in a tensor.
+        /// This method does not support jagged arrays since those require copies.
+        /// </summary>
+        /// <typeparam name="T">.NET type of tensor to create</typeparam>
+        /// <param name="data">value of tensor</param>
+        /// <param name="shape">shape of tensor. Only valid when passing in arrays.</param>
+        public static TFTensor Create<T>(T data, TFShape shape = null)
```

Comment:
>shape [](start = 57, length = 5)

Shape should be calculatable from Array, we shouldn't need to pass it down.  Are you doing this because you're trying to treat an array as a different shape tensor than its managed array's shape?

Reply:
Yes, data is kept in flatten array as row major format while shape indicates the dimensions. I was thinking to make a different create for arrays. What do you think?

---
In reply to: [210640853](https://github.com/zeahmed/machinelearning/pull/3#discussion_r210640853) [](ancestors = 210640853)

Reply:
I see.  A different method makes sense.  Can you call it CreateReshape or something like that?  It can live parallel to Create and could take T[] and avoid the typeof(T).IsArray/GetElementType code.

As an aside: does it cause issue that ML.NET forces n-dimensional data down into 1-d row-major layout?  That constraint feels limiting and seems like it would lead to extra work when bringing together components (end-user apps, transforms, etc) that need n-d or different layouts.

---
In reply to: [210698485](https://github.com/zeahmed/machinelearning/pull/3#discussion_r210698485) [](ancestors = 210698485,210640853)